### PR TITLE
fix: Numerous small performance and correctness issues

### PR DESCRIPTION
**Pull request by dpcollins-google**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-pubsublite/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: google/cloud/pubsublite/cloudpubsub/internal/client_multiplexer.py [hunk 1]
```diff
@@ -16,40 +16,33 @@
...
 
-from google.api_core.exceptions import FailedPrecondition
-
 _Key = TypeVar("_Key")
 _Client = TypeVar("_Client")
 
 
 class ClientMultiplexer(Generic[_Key, _Client]):
-    _OpenedClientFactory = Callable[[], _Client]
+    _OpenedClientFactory = Callable[[_Key], _Client]
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/cloudpubsub/internal/client_multiplexer.py:24]: There was a lot of "self" time in get_or_create callers from constructing the factory to pass in. This is in the publish hot path, hence this change

File: google/cloud/pubsublite/cloudpubsub/internal/client_multiplexer.py [hunk 2]
```diff
@@ -75,52 +68,45 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
 class AsyncClientMultiplexer(Generic[_Key, _Client]):
-    _OpenedClientFactory = Callable[[], Awaitable[_Client]]
+    _OpenedClientFactory = Callable[[_Key], Awaitable[_Client]]
     _ClientCloser = Callable[[_Client], Awaitable[None]]
 
+    _factory: _OpenedClientFactory
     _closer: _ClientCloser
-    _lock: asyncio.Lock
-    _live_clients: Dict[_Key, _Client]
+    _live_clients: Dict[_Key, Awaitable[_Client]]
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/cloudpubsub/internal/client_multiplexer.py:76]: There was a lot of time in acquiring the lock in the publish hotpath- this change makes the lock not necessary

File: google/cloud/pubsublite/cloudpubsub/internal/multiplexed_subscriber_client.py [hunk 3]
```diff
@@ -40,22 +38,16 @@ class MultiplexedSubscriberClient(SubscriberClientInterface):
     _executor: ThreadPoolExecutor
     _underlying_factory: AsyncSubscriberFactory
 
-    _multiplexer: ClientMultiplexer[SubscriptionPath, StreamingPullFuture]
+    _lock: Lock
+    _live_clients: Set[StreamingPullFuture]
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/cloudpubsub/internal/multiplexed_subscriber_client.py:42]: Previously, the subscriber enforced that there was only one open subscription stream per-subscription per-client, but there's actually no need for this.

File: google/cloud/pubsublite/internal/wire/serial_batcher.py [hunk 4]
```diff
@@ -54,12 +72,14 @@ def add(self, request: Request) -> "asyncio.Future[Response]":
         """
         item = WorkItem[Request, Response](request)
         self._requests.append(item)
+        self._batch_size += self._sizer.get_size(request)
         return item.response_future
 
-    def should_flush(self) -> bool:
-        return self._tester.test(item.request for item in self._requests)
+    def size(self) -> BatchSize:
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/wire/serial_batcher.py:78]: Most of the CPU in the publish path was in should_flush calls, which had to construct an iterable and iterate through every time.

File: google/cloud/pubsublite/internal/wire/single_partition_publisher.py [hunk 5]
```diff
@@ -189,10 +191,14 @@ async def reinitialize(
...
-            request_count += 1
-            byte_count += PubSubMessage.pb(req).ByteSize()
-        return (request_count >= _MAX_MESSAGES) or (byte_count >= _MAX_BYTES)
+    @overrides
+    def get_size(self, request: PubSubMessage) -> BatchSize:
+        return BatchSize(
+            element_count=1, byte_count=PubSubMessage.pb(request).ByteSize()
+        )
+
+    def _should_flush(self) -> bool:
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/wire/single_partition_publisher.py:200]: if you'll note on the left... batching settings were previously ignored

**Review by manuelmenzella-google — APPROVED** [review]

**Pull request merged**